### PR TITLE
high altitude description of disputes

### PR DESCRIPTION
**Pull request by drahnr**
After a discussion with @rphmeier , that #2192 might be doable but it might be preferable to do different impl details, this is a top down instead of bottom up spec of disputes.

### CODE REVIEW

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (burdges) [inline roadmap/implementers-guide/src/runtime/disputes2.md:13]: It's optional if we count their second vote, but we cannot permit later votes to replace their earlier votes of any voting type (dispute, approval, or backing), so it's almost surely simplest to allow voters to disagree with themselves.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (burdges) [inline roadmap/implementers-guide/src/runtime/disputes2.md:7]: It's technically possible two backers disagree too.  We should not waste too much time on this special case because a backing voter could create the dispute exactly like an approval checker can, but then availability should run concurrently with the dispute, and the disputing backer should continue handing out pieces, just like if they were a backing vote.  

An incorrect disputer need not be slashed 100% but they should be slashed enough that censoring a parachain by disputing blocks you dislike becomes prohibitively expensive, and correlated slashing rule apply to drive this upwards towards 100%.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (burdges) [inline roadmap/implementers-guide/src/runtime/disputes2.md:21]: "check the block" could "reconstruct and check the block" but maybe that's implicit based upon everything else

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (burdges) [inline roadmap/implementers-guide/src/runtime/disputes2.md:42]: I'd say the dispute should contain enough data to prove that all validators should begin processing the dispute, meaning at least one backing vote and at least one disputing vote.  

We're resolving using the validator set at inclusion time, so the backing vote could actually come from another validator set, but that's okay because the backers' stake remains locked and they remain liable.

**Comment by burdges** [comment]
I suppose some of my comments above are a bit too detailed for a high altitude description, so take them or leave them.  :)

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (drahnr) [reply roadmap/implementers-guide/src/runtime/disputes2.md:13]: But why would we allow a voter to contradict themselves? Isn't that already bad behaviour? If we would accept multiple votes of a validator, that would influence their weight. The simplest approach would be, to allow exactly one kind of vote per validator, and silently ignore vote types that are on the same side per validator (multiple plus votes are ok, but only the first one is accounted for), voting in differently with different vote types should be an attempted double vote.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (drahnr) [reply roadmap/implementers-guide/src/runtime/disputes2.md:42]: What would happen if noone is backing the proposed block?

Comment (burdges) [reply roadmap/implementers-guide/src/runtime/disputes2.md:40]: Good point!  If you receive an invalid candidate from a collator then what should you do?   It's not a dispute because no validator voted for the candidate yet.  It's almost surely buggy parachain code or buggy collator deployment, but maybe someone has launched an attack and just sent the bad candidate to you by mistake?  

There are two solutions, either (a) we hang onto the candidate until another candidate from that parachain gets finalized, or else (b) we throw it away and figure we can reconstruct ourselves if required, or else (c) ignore the problem and leave it to approval checkers.

I suspect Rob might dislike (a) so maybe we could implement (b) saying "backers have the right to reconstruct a block like tranche zero approval checkers".  In this way, you simply ignore invalid blocks that nobody else backs when backing, but then recheck under this bizarre situation.  We then later partially do (a) by saying that we deprioritize storing the block but do not actually flush out its storage yet, but this adds nothing formally. 

In the short term, I'd vote (c) ignore unbacked invalid blocks unless they're backed, but later maybe we'll try (b) or even this partial (a).

**Comment by drahnr** [comment]
Not sure how much detail can be added without dipping toes into a particular impl direction.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (burdges) [inline roadmap/implementers-guide/src/runtime/disputes2.md:17]: A priori, we do not require special punishment here.  Imagine someone votes both ways:  If the block turns out valid then they're slashed maybe 10%.  If otoh the block turns out invalid then they're slashed 100%.  If the block turns out indecisive, meaning no supermajority either way, then we lack the supermajority for any slash.  It's never in their interest to vote both ways because doing so only makes it more likely to reach a supermajority

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (rphmeier) [inline roadmap/implementers-guide/src/runtime/disputes2.md:5]: It's not clear what "the transaction" is here.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (burdges) [reply roadmap/implementers-guide/src/runtime/disputes2.md]: We cannot ignore votes if we cannot know which vote came first.  As I said below, we'll just count all votes because this harms the guy who votes multiple ways even without special handling.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (rphmeier) [reply roadmap/implementers-guide/src/runtime/disputes2.md:5]: Also, there is no dispute if everyone agrees the block is bad.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (rphmeier) [inline roadmap/implementers-guide/src/runtime/disputes2.md:9]: dispute implies conflict IMO. If there's nobody to slash because a backer just rejected some bad collator's block, then we have nobody to slash anyway.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (rphmeier) [inline roadmap/implementers-guide/src/runtime/disputes2.md:11]: The validator set needs to be trusted but the backing validators may not be.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (rphmeier) [inline roadmap/implementers-guide/src/runtime/disputes2.md:35]:
```suggestion
In order to ensure, the dispute result is not forgotten or intentionally side stepped, it has to be recorded on chain.
```

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (rphmeier) [inline roadmap/implementers-guide/src/runtime/disputes2.md:27]:
```suggestion
After an approval checker challenged a block, all validators that received the gossiped vote, reconstruct the block
```

Generally I think we should be circulating a backing vote with it.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (rphmeier) [reply roadmap/implementers-guide/src/runtime/disputes2.md:45]: I think we've got to accept disputes on anything that at least one validator has issued a backing statement for. Otherwise we'd have to mess around with trying to prove to some remote chain whether a block was actually backed by a majority of its backing validators, and who those validators were, etc.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (rphmeier) [inline roadmap/implementers-guide/src/runtime/disputes2.md:55]: I think we should avoid voting to finalize any chains that have an active dispute or have had a dispute resolve to the negative. This is combined with the voting rule of not finalizing anything which doesn't appear to have been approved by enough approval checkers.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (rphmeier) [reply roadmap/implementers-guide/src/runtime/disputes2.md:45]: Well, now that I think about it, proving this to a remote chain might not actually be that hard assuming that every chain has the same view of sessions / validators / paras. But that is kind of a stretch as we don't fix paras one session in advance the same way we do validators/randomness. It is a change I am open to, though.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (drahnr) [reply roadmap/implementers-guide/src/runtime/disputes2.md:17]: This works due to our base assumption that less than ⅓ of validators are malicious, which implies if all malicious validators vote both ways, we can never reach a supermajority (or multiple supermajorities) in any case.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (drahnr) [reply roadmap/implementers-guide/src/runtime/disputes2.md:27]: How do we pick the backing vote? Or should all backing votes be circulated?

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (drahnr) [reply roadmap/implementers-guide/src/runtime/disputes2.md:45]: But the implications of noone backing that particular block would be no inclusion, as such why would we care to deal with it any further down the row, instead of just dropping the negative vote? I think we need to talk some more about the details of that.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (drahnr) [reply roadmap/implementers-guide/src/runtime/disputes2.md]: How do we deal with multple votes that are both pro (backing + explicit pro vote), where one comes during the open dispute, but the second comes in after dispute resolution. Then we would need to either reward the full or the reduced amount, imho it's the earlier.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (burdges) [reply roadmap/implementers-guide/src/runtime/disputes2.md:45]: Actually my (b) sounds pointless now, so our approach goes roughly:

We've a backer who obtains a candidate B but discovers B to be invalid.  Initially, our backer does nothing but ideally^2 retains B for a little while.  If our backer ever sees a backing vote for B then our backer should ideally^1 create a dispute, and retain B permanently if they have B.  Ideally^3 our disputing backer should then participate in erasure coded piece distribution.  Ideally^4 if a backer appears absent then we ask for one more approval checker, but do not count the missing backer as a no-show.

We've three simplifications obtained by removing bits of functionality prefixed by ideally^i for i>k.  All four ideally^i add only small-ish improvements so I do not want to push for them when people have higher priority things.  If an ideally^i looks really easy to do, then go for it.  :)  If we add none of these ideally^i then backers simply ignore invalid candidates.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (ordian) [inline roadmap/implementers-guide/src/runtime/disputes2.md:19]: It would be helpful to define the terms `quorum` and `supermajority`.
```suggestion
advance the progress towards supermajority either way and have their bonds slashed.
```

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (ordian) [inline roadmap/implementers-guide/src/runtime/disputes2.md:63]: TBD?

**Review by ordian — APPROVED** [review]
Looks good in general modulo unresolved comments.

We also probably want to rename the file and place it somewhere in the table of contents.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (drahnr) [reply roadmap/implementers-guide/src/runtime/disputes2.md:63]: To be decided, it's a bit out of scope for this document as well, since the logic of dispute resolution should not be influenced by the fork-choice rule.

File: roadmap/implementers-guide/src/protocol-disputes.md
Comment (rphmeier) [inline roadmap/implementers-guide/src/protocol-disputes.md:17]:
```suggestion
The existence of a dispute is detected by a backing checker
```

File: roadmap/implementers-guide/src/protocol-disputes.md
Comment (rphmeier) [inline roadmap/implementers-guide/src/protocol-disputes.md:12]: usually called Reed-Solomon

Comment (rphmeier) [reply roadmap/implementers-guide/src/protocol-disputes.md:12]:
```suggestion
Now the dispute can be happening quite some time later than the inclusion, but also right during backing. As such the block is stored or more so its Reed-Solomon encoded erasure chunks, from which the
```

File: roadmap/implementers-guide/src/protocol-disputes.md
Comment (rphmeier) [inline roadmap/implementers-guide/src/protocol-disputes.md:38]:
```suggestion
A dispute is initiated by one approval checker creating and gossiping a vote, that challenges another vote.
```

For clarity

File: roadmap/implementers-guide/src/protocol-disputes.md
Comment (rphmeier) [inline roadmap/implementers-guide/src/protocol-disputes.md:39]: This sentence is a little out-of-place

Also, better to just refer to blocks as opposed to transactions

File: roadmap/implementers-guide/src/protocol-disputes.md
Comment (rphmeier) [inline roadmap/implementers-guide/src/protocol-disputes.md:41]: Previously the document mentions that backing checkers can also initiate disputes. This section implies that it is the sole responsibility of approval checkers.

File: roadmap/implementers-guide/src/protocol-disputes.md
Comment (rphmeier) [inline roadmap/implementers-guide/src/protocol-disputes.md:51]: I agree with everything except for the 'iff' - we don't want malicious validators to be able to avoid slashing just because they reorganize the chain to no longer contain their attempt at including a bad parablock.

File: roadmap/implementers-guide/src/protocol-disputes.md
Comment (rphmeier) [inline roadmap/implementers-guide/src/protocol-disputes.md:54]: I would typically phrase this as the "chain" must be put in governance mode

File: roadmap/implementers-guide/src/protocol-disputes.md
Comment (drahnr) [reply roadmap/implementers-guide/src/protocol-disputes.md:12]: Don't ask why this is stuck the other way round in my head...

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (drahnr) [reply roadmap/implementers-guide/src/runtime/disputes2.md]: Not quite sure this section fits in here, or if that should be covered in a document itself and just be linked. It seems a bit out of scope for disputes and not affect the actual logic at all.

File: roadmap/implementers-guide/src/runtime/disputes2.md
Comment (drahnr) [reply roadmap/implementers-guide/src/runtime/disputes2.md]: Ideally^1 & Ideally^2 seem reasonable to impl, not sure about ^3 and ^4

**Comment by drahnr** [comment]
bot merge

**Pull request merged**
